### PR TITLE
Restore old SubArray bounds check semantics

### PR DESCRIPTION
**Pull request by mbauman**
That is, don't check bounds and instead let the parent array do the bounds checking.  Due to the abstract fallbacks, it's still advantageous to define both `getindex` and `unsafe_getindex`.  It'd be nice to use metaprogramming here to loop over the two names and define both at once, but it seems like `subarray.jl` lives too early in the bootstrap for the required machinery.

One thing that might be nice to address before merging is the fact that this (and 0.4-before-10525) doesn't do any bounds checking when doing indexing lookup.  For `Range` that's just fine, since it's simply doing math, but when one of the SubArray's index vectors is an Array, this can lead to segfaults:

```
julia> sub([1:10;], [1,2])[10000000000]

signal (11): Segmentation fault: 11
anonymous at subarray.jl:543
jl_apply at /Users/mbauman/Code/julia-0.4/src/interpreter.c:55
eval at /Users/mbauman/Code/julia-0.4/src/interpreter.c:212
jl_toplevel_eval_flex at /Users/mbauman/Code/julia-0.4/src/toplevel.c:523
jl_toplevel_eval_in at /Users/mbauman/Code/julia-0.4/src/builtins.c:539
eval_user_input at REPL.jl:62
jlcall_eval_user_input_21081 at  (unknown line)
anonymous at task.jl:91
jl_apply at /Users/mbauman/Code/julia-0.4/src/task.c:234
Segmentation fault: 11
```

### CODE REVIEW

File: base/subarray.jl [hunk 1]
```diff
@@ -527,7 +527,22 @@ function parentdims(s::SubArray)
 end
 
 ## Scalar indexing
-getindex(V::SubArray, I::Int...) = (@_inline_meta; checkbounds(V, I...); unsafe_getindex(V, I...))
```
Comment (timholy) [inline base/subarray.jl]: We could presumably use a loop `for func in (:getindex, :unsafe_getindex)`? But I'm fine either way.

**Comment by timholy** [comment]
Thanks for tackling this! Yes, I too look forward to the day where we can revert this.

I agree that it would be nice to check bounds on the index vectors, but I suspect that's another case where we need better `@inbounds` technology. I think I tried having it when I was developing the new SubArrays, and at least at the time the performance hit was large. All this is basically why I've long advanced the position that we can't make `A[2:n, 7]` return a SubArray until we have bounds-checking solved---it has to be opt-in (via calling `sub` or `slice`), or we'd lose too much safety.

File: base/subarray.jl (hunk 1 above)
Comment (mbauman) [inline base/subarray.jl]: Yup, that's what I tried first, but this loads before `array.jl` in the sysimg, so the machinery to do that isn't available yet.

**Comment by timholy** [comment]
I would like to hit merge on this, but haven't had time to keep up with the latest on "spurious" test failures. Are these anything to worry about?

**Comment by yuyichao** [comment]
The Win64 failure looks new to me (but I may be wrong)

**Comment by mbauman** [comment]
I'm not sure, either.  I figured I'd add a comment in the code to explain why we're doing it this way, and push again to trigger a new build.  For posterity, the strange segfault was in libgmp (_gmpn_mul_1 ) without any sensible backtrace: https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.5545/job/y5ib6qpr4e423f4w

I took a quick look and I agree, checking bounds on the indexing vectors can be just as costly.  Perhaps we should document that indexing into subarrays is unchecked?

**Comment by timholy** [comment]
Yeah, we should do better about documenting that.

**Comment by tkelman** [comment]
The win64 failure looks new to me as well. I believe there is an expected sigfpe that comes from gmp during the numbers test, but it's supposed to be caught by Julia's signal handlers. I wonder if/how this change could modify that in any way? Or if something recently changed on master that could lead to that but intermittently.

**Comment by mbauman** [comment]
Alright, well, this has now passed through the Appveyor minefield twice unscathed.  As much as I'd hate to add yet another spurious failure, I don't think that this change is the root of that segfault since it should (at worst) only cause an out-of-bounds error that gets caught by --check-bounds=yes.  I'll merge it.

**Comment by timholy** [comment]
I agree---there's no reason this by itself should be problematic. Thanks!

**Pull request merged**